sm/download/src/download.rs

```diff
@@ -0,0 +1,94 @@
+use crate::error::DownloadError;
+use backoff::{future::retry, ExponentialBackoff};
+use json_sm::DownloadInfo;
+use log::error;
+use std::{
+    path::{Path, PathBuf},
+    time::Duration,
+};
+
+#[derive(Debug)]
+pub struct Downloader {
+    target_filename: PathBuf,
+    download_target: PathBuf,
+}
+
+impl Downloader {
+    pub fn new(name: &str, version: &Option<String>, target_dir_path: impl AsRef<Path>) -> Self {
+        let mut filename = name.to_string();
+        if let Some(version) = version {
+            filename.push('_');
+            filename.push_str(version.as_str());
+        }
+
+        let mut download_target = PathBuf::new().join(&target_dir_path).join(&filename);
+        download_target.set_extension("tmp");
+        let target_filename = PathBuf::new().join(target_dir_path).join(filename);
+
+        Self {
+            target_filename,
+            download_target,
+        }
+    }
+
+    pub async fn download(&self, url: &DownloadInfo) -> Result<(), DownloadError> {
+        // Default retry is an exponential retry with a limit of 15 minutes total.
+        // Let's set some more reasonable retry policy so we don't block the downloads for too long.
+        let backoff = ExponentialBackoff {
+            initial_interval: Duration::from_secs(30),
+            max_elapsed_time: Some(Duration::from_secs(300)),
+            ..Default::default()
+        };
+
+        let response = retry(backoff, || async {
+            let client = reqwest::Client::new();
+            match &url.auth {
+                Some(json_sm::Auth::Bearer(token)) => {
+                    match client
+                        .get(url.url())
+                        .bearer_auth(token)
+                        .send()
+                        .await
+                        .unwrap()
+                        .error_for_status()
+                    {
+                        Ok(response) => Ok(response),
+                        Err(err) => {
+                            error!("Request returned an error: {:?}", &err);
+                            Err(err.into())
+                        }
+                    }
+                }
+                None => match client.get(url.url()).send().await?.error_for_status() {
+                    Ok(response) => Ok(response),
+                    Err(err) => {
+                        error!("Request returned an error: {:?}", &err);
+                        Err(err.into())
+                    }
+                },
+            }
+        })
+        .await?;
+
+        let content = response.bytes().await?;
```

Comment:
We cannot assume that the full content of the file can be read in memory.

That is only okay for a very first version of that crate.